### DISCUSSION TITLE
What is the advantage of the packed sequences?

### DISCUSSION THREAD

**Discussion by TranstecJasonD** [NetTopologySuite/NetTopologySuite#475] [Q&A] [answered]
Background: I need a geometry that is the result of the intersection of two other geometries. Using linear rings for the geometries does not work because the intersection returns an unordered Multipoint object. So I am using Polygons.

Everything works great but I wanted to use packed arrays to avoid creating arrays of Coordinate objects. Creating the geometries this way works fine. However when I call Intersection, Coordinate objects and arrays are created by the function. In fact, comparing the allocations of packed vs. non-packed, the non-packed uses less memory and fewer allocations (checked using the Visual Studio profiler). So what is the advantage of using packed if the Coordinate objects are just going to be created anyway? Is it more of a convenience if your data already happens to be in that form? I realize this is just one scenario so maybe other scenarios are different. While it is certainly not a huge issue, I am rewriting my NTS code anyway and would like to choose the best option for me and get a better understanding.

**Comment by airbreather** [comment, marked as answer]
Great question!

When you are storing a `Geometry` instance that will be kept alive for a long time, the packed sequences are your friend: holding a separate `Coordinate` instance for each point costs roughly 16 bytes of overhead per point (in a 64-bit process, anyway) and, potentially, heap fragmentation.

There are some algorithms (I'm not in a position to list any at this exact moment) that should operate just as efficiently (if not more efficiently) when using packed sequences like this.

Additionally, packed sequences can make it cheaper to interoperate with other libraries / frameworks that expect to see data laid out in this form, and/or make it easier to save / load instances by copying directly from some buffer into the final target arrays.

The main cost of this is what you've observed: many algorithms aren't tuned to work directly off of the `CoordinateSequence`, instead choosing to allocate a bunch of `Coordinate` instances for their own purposes.  Since the packed sequences don't already have these `Coordinate` instances, they will have to be created and allocated on-the-fly, which will almost certainly slow down those operations.

**Reply by TranstecJasonD** [reply]
Thanks for the prompt response. That helps a lot. The other library framework is one of the main reasons I'm looking into this. I'm migrating from DotSpatial to PROJ and the latter has the ability to use packed arrays so I was looking to use the same arrays for both libraries when possible.
